File: Flutter-client/lib/common/api/game_order/game_order_api.dart
// ignore_for_file: unused_element

import 'package:gogaming_app/widget_header.dart';

import '../base/base_api.dart';

enum GameType {
  sportsBook('SportsBook'),
  lottery('Lottery'),
  casino('Casino'),
  chess('Chess');

  const GameType(this.value);
  final String value;
}

extension GameTypeExtension on GameType {
  static GameType convert(String? value) {
    switch (value) {
      case 'SportsBook':
        return GameType.sportsBook;
      case 'Lottery':
        return GameType.lottery;
      case 'Casino':
        return GameType.casino;
      case 'Chess':
        return GameType.chess;
      default:
        return GameType.sportsBook;
    }
  }

  String get translate {
    switch (this) {
      case GameType.sportsBook:
        return localized('sports');
      case GameType.lottery:
        return localized('lottery');
      case GameType.casino:
        return localized('casino');
      case GameType.chess:
        return localized('chess');
    }
  }
}

enum GameOrder with GoGamingApi {
  /// 彩票游戏订单详情
  getLotteryDetail(params: {"wagerNumber": "交易单号"}),

  /// 体育游戏订单详情
  getSportDetail(params: {"wagerNumber": "交易单号"}),

  /// 棋牌游戏订单详情
  getChessDetail(params: {"wagerNumber": "交易单号"}),

  /// 娱乐场游戏订单详情
  getCasinoDetail(params: {"wagerNumber": "交易单号"}),

  recentOrder(params: {
    'pageIndex': 1,
    'pageSize': 9,
  }),

  /// 注单日期汇总，指定时间段内，按日期统计返回注单数量
  ///
  /// [beginDate] 和 [endDate] 需采用UTC+时差的时间
  ///
  /// 建议使用GamingDateRangeSelector.show的isUtc进行日期选择
  getWagerDayTotal(params: {
    'gameType': 'x',
    'wagerStatus': null,
    'beginDate': 0,
    'endDate': 0,
  }),

  /// 按日期获取注单列表
  ///
  /// [beginTime] getWagerDayTotal返回的日期+时差
  getWagerList(params: {
    'gameType': 'x',
    'wagerStatus': null,
    'beginTime': 0,
    'endTime': 0,
    'page': 1,
    'pageSize': 10,
  }),

  /// 我的投注初始化
  getSelfRealTimeBetInfo(params: {'pageSize': 10}),

  /// 所有投注初始化
  getRealTimeBetInfo(params: {'pageSize': 10}),

  /// 风云榜
  getHeroBetInfo(params: {'pageSize': 10}),

  /// 最幸运
  getLuckiestBetInfo(params: {'pageSize': 10}),

  /// 大赢家
  getMostWinerBetInfo(
      params: {'gameId': '游戏id', 'provider': '游戏厂商', 'catId': '游戏类型'}),

  /// 幸运赢家
  getLuckyWinerBetInfo(
      params: {'gameId': '游戏id', 'provider': '游戏厂商', 'catId': '游戏类型'}),

  getWagerStatusSelect(params: {
    'gameType': 'x',
  });

  const GameOrder({this.params, this.data});

  @override
  final Map<String, dynamic>? params;
  @override
  final Map<String, dynamic>? data;

  @override
  String get path {
    switch (this) {
      case GameOrder.recentOrder:
        return '/asset/gameorder/recentorder';
      case GameOrder.getLotteryDetail:
        return '/asset/gameorder/getlotterydetail';
      case GameOrder.getSportDetail:
        return '/asset/gameorder/getsportdetail';
      case GameOrder.getChessDetail:
        return '/asset/gameorder/getchessdetail';
      case GameOrder.getCasinoDetail:
        return '/asset/gameorder/getcasinodetail';
      case GameOrder.getWagerList:
        return '/asset/gameorder/getwagerlist';
      case GameOrder.getWagerDayTotal:
        return '/asset/gameorder/getwagerdaytotal';
      case GameOrder.getWagerStatusSelect:
        return '/asset/gameorder/getwagerstatusselect';
      case GameOrder.getSelfRealTimeBetInfo:
        return '/asset/gameorder/getselfrealtimebetinfo';
      case GameOrder.getRealTimeBetInfo:
        return '/asset/gameorder/getrealtimebetinfo';
      case GameOrder.getHeroBetInfo:
        return '/asset/gameorder/getherobetinfo';
      case GameOrder.getLuckiestBetInfo:
        return '/asset/gameorder/getluckiestbetinfo';
      case GameOrder.getMostWinerBetInfo:
        return '/asset/gameorder/getmostwinerbetinfo';
      case GameOrder.getLuckyWinerBetInfo:
        return '/asset/gameorder/getluckywinerbetinfo';
    }
  }

  @override
  HTTPMethod get method {
    switch (this) {
      case GameOrder.recentOrder:
      case GameOrder.getLotteryDetail:
      case GameOrder.getSportDetail:
      case GameOrder.getChessDetail:
      case GameOrder.getCasinoDetail:
      case GameOrder.getWagerList:
      case GameOrder.getWagerDayTotal:
      case GameOrder.getWagerStatusSelect:
      case GameOrder.getSelfRealTimeBetInfo:
      case GameOrder.getRealTimeBetInfo:
      case GameOrder.getHeroBetInfo:
      case GameOrder.getLuckiestBetInfo:
      case GameOrder.getMostWinerBetInfo:
      case GameOrder.getLuckyWinerBetInfo:
        return HTTPMethod.get;
    }
  }
}

File: Client-side/src/app/shared/apis/gameorder.api.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from 'src/environments/environment';
import {
  BigWinnerParams,
  CasinoDealRecord,
  CasinoDealRecordDetail,
  ChessDealRecord,
  ChessDealRecordDetail,
  CommonRealTimeData,
  DealRecordParams,
  GetBigWinner,
  GetLuckiestUser,
  LotteryDealRecord,
  LotteryDealRecordDetail,
  LuckiestUserParams,
  RecentorderDealRecordParams,
  RecentTransactionsData,
  SportDealRecord,
  SportDealRecordDetail,
  StatusSelect,
  WagerDaysTotalData,
} from '../interfaces/gameorder.interface';
import { ResponseData, ResponseListData } from '../interfaces/response.interface';
import { BaseApi } from './base.api';

@Injectable({
  providedIn: 'root',
})
export class GameOrderApi extends BaseApi {
  private get gameorderUrl(): string {
    return `${environment.apiUrl}/v1/asset/gameorder`;
  }

  /**
   * 查询订单状态下拉列表
   *
   * @param gameType
   */
  getStatusSelect(gameType: 'SportsBook' | 'Lottery' | 'Casino' | 'Chess'): Observable<ResponseData<StatusSelect[]>> {
    return this.get(`${this.gameorderUrl}/getwagerstatusselect`, { gameType });
  }

  /**
   * 查询【体育】交易记录列表
   * 返回 SportDealRecord[]
   *
   * @param params
   */
  getsportlist(params: DealRecordParams): Observable<ResponseListData<SportDealRecord[]>> {
    return this.get(`${this.gameorderUrl}/getsportlist`, params);
  }

  /**
   * 查询【体育】交易记录详情
   * wagerNumber 交易单号
   * 返回 SportDealRecordDetail[]
   *
   * @param wagerNumber
   */
  getsportdetail(wagerNumber: string): Observable<ResponseData<SportDealRecordDetail[]>> {
    return this.get(`${this.gameorderUrl}/getsportdetail`, { wagerNumber });
  }

  /**
   * 查询【娱乐城】交易记录列表
   * 返回 CasinoDealRecord[]
   *
   * @param params
   */
  getcasinolist(params: DealRecordParams): Observable<ResponseListData<CasinoDealRecord[]>> {
    return this.get(`${this.gameorderUrl}/getcasinolist`, params);
  }

  /**
   * 查询【娱乐城】交易记录详情
   * wagerNumber 交易单号
   * 返回 CasinoDealRecordDetail
   *
   * @param wagerNumber
   */
  getcasinodetail(wagerNumber: string): Observable<ResponseData<CasinoDealRecordDetail>> {
    return this.get(`${this.gameorderUrl}/getcasinodetail`, { wagerNumber });
  }

  /**
   * 查询【棋牌】交易记录列表
   * 返回 ChessDealRecord[]
   *
   * @param params
   */
  getchesslist(params: DealRecordParams): Observable<ResponseListData<ChessDealRecord[]>> {
    return this.get(`${this.gameorderUrl}/getchesslist`, params);
  }

  /**
   * 查询【棋牌】交易记录详情
   * wagerNumber 交易单号
   * 返回 ChessDealRecordDetail
   *
   * @param wagerNumber
   */
  getchessdetail(wagerNumber: string): Observable<ResponseData<ChessDealRecordDetail>> {
    return this.get(`${this.gameorderUrl}/getchessdetail`, { wagerNumber });
  }

  /**
   * 查询【彩票】交易记录列表
   * 返回 LotteryDealRecord[]
   *
   * @param params
   */
  getlotterylist(params: DealRecordParams): Observable<ResponseListData<LotteryDealRecord[]>> {
    return this.get(`${this.gameorderUrl}/getlotterylist`, params);
  }

  /**
   * 查询【彩票】交易记录详情
   * wagerNumber 交易单号
   * 返回 LotteryDealRecordDetail
   *
   * @param wagerNumber
   */
  getlotterydetail(wagerNumber: string): Observable<ResponseData<LotteryDealRecordDetail>> {
    return this.get(`${this.gameorderUrl}/getlotterydetail`, { wagerNumber });
  }

  /**
   * 查询【最近交易】记录详情
   * 返回 RecentorderDealRecord
   *
   * @param params
   */
  getRecentorder(params: RecentorderDealRecordParams): Observable<ResponseListData<RecentTransactionsData[]>> {
    return this.get(`${this.gameorderUrl}/recentorder`, params);
  }

  /**
   * 查询会员注单日期汇总 TODO: 时间还原
   *
   * @param gameType 类别
   * @param wagerStatus 状态
   * @param beginDate 开始时间
   * @param endDate 结束时间
   * @returns
   */
  getWagerDaytotal(
    gameType: 'SportsBook' | 'Lottery' | 'Casino' | 'Chess',
    wagerStatus: string,
    beginDate: number,
    endDate: number
  ): Observable<ResponseData<WagerDaysTotalData>> {
    return this.get(`${this.gameorderUrl}/getwagerdaytotal`, { gameType, wagerStatus, beginDate, endDate });
  }

  /**
   * 统一注单列表
   *
   * @param gameType 类别
   * @param wagerStatus 状态
   * @param beginTime 开始时间
   * @param endTime 结束时间
   * @param page 页码
   * @param pageSize 每页大小
   * @returns
   */
  getWagerList(
    gameType: 'SportsBook' | 'Lottery' | 'Casino' | 'Chess',
    wagerStatus: string,
    beginTime: number,
    endTime: number,
    page: number,
    pageSize: number
  ): Observable<ResponseListData<CasinoDealRecord[] | SportDealRecord[] | ChessDealRecord[] | LotteryDealRecord[]>> {
    return this.get(`${this.gameorderUrl}/getwagerlist`, { gameType, wagerStatus, beginTime, endTime, page, pageSize });
  }

  /**
   * 我的投注初始化
   *
   * @param pageSize 当前页面size
   * @param gameCategories
   * @returns
   */
  getMyBet(pageSize: number, gameCategories: string[]): Observable<Array<CommonRealTimeData>> {
    return this.get<ResponseData<Array<CommonRealTimeData>>>(`${this.gameorderUrl}/getselfrealtimebetinfo`, {
      pageSize,
      gameCategories,
    }).pipe(map(v => v?.data || []));
  }

  /**
   * 所有投注
   *
   * @param pageSize 当前页面size
   * @param gameCategories
   * @returns
   */
  getAllBets(pageSize: number, gameCategories: string[]): Observable<Array<CommonRealTimeData>> {
    return this.get<ResponseData<Array<CommonRealTimeData>>>(`${this.gameorderUrl}/getrealtimebetinfo`, {
      pageSize,
      gameCategories,
    }).pipe(map(v => v?.data || []));
  }

  /**
   * 风云榜
   *
   * @param pageSize 当前页面size
   * @param gameCategories
   * @returns
   */
  getHeroBet(pageSize: number, gameCategories: string[]): Observable<Array<CommonRealTimeData>> {
    return this.get<ResponseData<Array<CommonRealTimeData>>>(`${this.gameorderUrl}/getherobetinfo`, {
      pageSize,
      gameCategories,
    }).pipe(map(v => v?.data || []));
  }

  /**
   * 最幸运
   *
   * @param pageSize
   * @param gameCategories
   * @returns
   */
  getLuckiest(pageSize: number, gameCategories: string[]): Observable<Array<CommonRealTimeData>> {
    return this.get<ResponseData<Array<CommonRealTimeData>>>(`${this.gameorderUrl}/getluckiestbetinfo`, {
      pageSize,
      gameCategories,
    }).pipe(map(v => v?.data || []));
  }

  /**
   * 大隐家
   *
   * @param params
   * @returns
   */
  getBigWinner(params: BigWinnerParams): Observable<Array<GetBigWinner>> {
    return this.get<ResponseData<Array<GetBigWinner>>>(`${this.gameorderUrl}/getmostwinerbetinfo`, params).pipe(
      map(v => v?.data || [])
    );
  }

  /**
   * 幸运玩家
   *
   * @param params
   * @returns
   */
  getLuckiestUser(params: LuckiestUserParams): Observable<Array<GetLuckiestUser>> {
    return this.get<ResponseData<Array<GetLuckiestUser>>>(`${this.gameorderUrl}/getluckywinerbetinfo`, params).pipe(
      map(v => v?.data || [])
    );
  }
}
